### PR TITLE
Add servant 0.19 support to servant-util

### PR DESCRIPTION
**Pull request by marinelli**
This PR proposes to:
- Add support to build with `servant` >= v0.19 (fix https://github.com/serokell/servant-util/issues/38)
- Add `cabal.project` file
- Fix building with `universum-1.7.3`

#### Related changes (conditional)

- [ ] Tests
  - If I added new functionality, I added tests covering it.
  - If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- [ ] Documentation
  - I checked whether I should update the docs and did so if necessary:
    - [README](/README.md);
    - Haddock;
    - [Examples](/examples/).

- [ ] Public contracts
  - Any modifications of public contracts comply with the [Evolution
  of Public Contracts](https://www.notion.so/serokell/Evolution-of-Public-Contracts-2a3bf7971abe4806a24f63c84e7076c5) policy.
  - I added an entry to the [changelog](../tree/master/CHANGES.md) if my changes are visible to the users
        and
  - provided a migration guide for breaking changes if possible.

#### Stylistic guide (mandatory)

- [ ] Style
  - My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
  - My code complies with the [style guide](../tree/master/docs/code-style.md).
  - Each commit of this pull request contains a description.
    - For new features, this description contains the use case for the new functionality.
    - For bug fixes, it describes both the attacked problem and a solution for it.
    - For dependency version changes description shortly enlists features and breaking changes of the new library version or contains a link to its changelog.

### CODE REVIEW

File: servant-util/src/Servant/Util/Combinators/Filtering/Base.hs [hunk 1]
```diff
@@ -90,6 +97,18 @@ unsupportedFilteringValue errMsg = FilteringValueParser (\_ -> Left errMsg)
...
+-- It's required due to this https://github.com/haskell-servant/servant/pull/1432
+#if MIN_VERSION_servant(0,19,0)
+type EncodedQueryParam = ByteString
+encodeQueryParam :: ToHttpApiData a => a  -> EncodedQueryParam
+encodeQueryParam = BL.toStrict . toLazyByteString . toEncodedUrlPiece
+#else
+type EncodedQueryParam = Text
+encodeQueryParam :: ToHttpApiData a => a  -> EncodedQueryParam
+encodeQueryParam = toQueryParam
+#endif
```
Comment (treeowl) [inline servant-util/src/Servant/Util/Combinators/Filtering/Base.hs:110]: This seems really awkward. How bad would it be to bump the dependency version?

Comment (marinelli) [reply servant-util/src/Servant/Util/Combinators/Filtering/Base.hs:110]: I do not like this solution too. If supporting servant 0.18 is not required, let's remove it.
Btw, I'll create an other PR to export `encodeQueryParam` from the `servant-client-core` package.

Comment (marinelli) [reply servant-util/src/Servant/Util/Combinators/Filtering/Base.hs]: Did you were thinking something like this https://github.com/marinelli/servant-util/commit/a0dbeeb24212e1a21e0fc332abc5523e5d939029?

Comment (marinelli) [reply servant-util/src/Servant/Util/Combinators/Filtering/Base.hs]: @Martoon-00 what do you think about it?

Comment (Martoon-00) [reply servant-util/src/Servant/Util/Combinators/Filtering/Base.hs]: I apologize for the late reply :pray: :pray: 

I don't like such tricks with CPP pragmas too, but in this case its scope is sufficiently limited and it seems like the least evil to me.

Let's go with the current solution to leave others some time if they need a recent feature of `servant-util` but cannot afford to switch to the `servant-0.19`. And after a couple of major releases of `servant`, or if this `CPP` turns out to hinder other features, we will strip the old behaviour.

I also asked our OPS team to make CI run both the old and new versions of `servant-client-core` ([internal ticket for this](https://issues.serokell.io/issue/OPS-1315)).

---

@marinelli You are also very welcome to start using https://github.com/haskell-servant/servant/pull/1549 here as soon as it appears in a release. Just leave a note in the changelog that we do not support `servant-0.19`, rather only `< 0.19` and `>= 0.19.X`.

Thanks a lot for your work!

File: servant-util/servant-util.cabal [hunk 2]
```diff
@@ -111,6 +111,7 @@ library
       QuickCheck
     , aeson
     , base >=4.7 && <5
+    , bytestring
```
[the remark this answers is not in the document]
Comment (marinelli) [reply servant-util/servant-util.cabal:114]: done

File: servant-util/src/Servant/Util/Combinators/Filtering/Base.hs (hunk 1 above)
Comment (marinelli) [reply servant-util/src/Servant/Util/Combinators/Filtering/Base.hs:110]: I'm not sure anymore that re-exporting `encodeQueryParamValue` from `servant-client-core` is a good idea :)
We'll see.

File: servant-util/src/Servant/Util/Combinators/Sorting/Construction.hs [hunk 3]
```diff
@@ -79,7 +80,7 @@ sortingSpec = mkSortingSpec [asc #id]
 mkSortingSpec
     :: ReifySortingItems base
     => [SortingRequestItem provided] -> SortingSpec provided base
-mkSortingSpec = fromList
+mkSortingSpec = GHC.Exts.fromList
```
Comment (Martoon-00) [inline servant-util/src/Servant/Util/Combinators/Sorting/Construction.hs:83]: Thanks :+1:

**Review by Martoon-00 — APPROVED** [review]
Again thank you for the update.

Please rebase and let's merge it.

**Comment by marinelli** [comment]
> Again thank you for the update.
> 
> Please rebase and let's merge it.

I rebased and updated the `servant-util/CHANGES.md` file.

**Pull request merged**